ENH: random: Add multivariate_hypergeometric function.

WarrenWeckesser opened the pull request:
Add a function that draws from a multivariate hypergeometric distribution.

WarrenWeckesser committed:
ENH: random: Add multivariate_hypergeometric function.

WarrenWeckesser commented:
A description of the distribution on wikipedia: https://en.wikipedia.org/wiki/Hypergeometric_distribution#Multivariate_hypergeometric_distribution

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
 
+    def _multivariate_hypergeometric1(self, ndarray colors,
+                                      npy_intp nsample, size=None):
+        # Algorithm 1 to sample from the multivariate hypergeometric
```
Why not a regular docstring

WarrenWeckesser replied:
For no good reason!  I'll change it.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
+        # To generate a sample using rk_hypergeometric:
+        #    red_sample = rk_hypergeometric(ngood=nred, nbad=total-nred, nsample=nsample)
+        # The gives us the number of red marbles in the sample.  The number of
+        # not red marbles in the sample is nsample - red_sample.  To figure out the distribution
```
Some of these lines are too long. We still try to keep them < 80 characters.

WarrenWeckesser replied:
Oops, I'll fix those.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
 
         return multin
 
+    def _multivariate_hypergeometric1(self, ndarray colors,
```
Is this a C function? Any reason that it isn't `cdef`?

WarrenWeckesser replied:
For testing and timing, I had exposed the two private functions at the end of `mtrand.pyx` as:

    multivariate_hypergeometric1 = _rand._multivariate_hypergeometric1
    multivariate_hypergeometric2 = _rand._multivariate_hypergeometric2

But since they aren't public in the version in the pull request, I'll change them to `cdef`.

charris replied:
Could then pass `colors` as a pointer. Might be easier to be sure of type consistency that way.

charris replied:
They don't need to be methods either, could just make them stand alone functions as long as they don't have internal state.

WarrenWeckesser replied:
They use `self.lock` and `self.internal_state`.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
+        # Example: Suppose the urn contains red, green, blue and yellow marbles.
+        # Let nred be the number of red marbles, and define the quantities for the
+        # other colors similarly.  The total number of marbles is
+        #    total =nred+ngreen+nblue+nyellow.
```
Can be better formatted using some spaces.

WarrenWeckesser replied:
Yup.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -639,6 +639,77 @@ def test_multivariate_normal(self):
+        size = (2, 3)
+        sample = np.random.multivariate_hypergeometric(colors, nsample, size)
+        assert_array_equal(sample.sum(axis=-1),
+                           np.full(size, fill_value=nsample, dtype=int))
```
I think you can use a simple scalar here instead of an array of `nsample`.

WarrenWeckesser replied:
I noticed that, but I didn't use it because it is not actually documented in the function's docstring, and I don't think it should behave that way.  If I expect a result to be [10, 10, 10], and I use `assert_array_equal(result, 10)`, it will (incorrectly) pass if `result` is 10.  Note that using a scalar value to check that an array contains all those values actually contradicts the docstring, which says "...check that the shape is equal...".  So I would say the fact that `assert_array_equal([10, 10, 10], 10)` doesn't raise an exception is a bug.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -639,6 +639,77 @@ def test_multivariate_normal(self):
+        # expected values.
+        assert_allclose(sample.mean(axis=0), nsample*colors/colors.sum(),
+                        rtol=1e-3, atol=0.005)
+
```
Another edge case would be setting `nsample = np.sum(colors)`

WarrenWeckesser replied:
There is such a test, but I'll add another that also uses the `size` argument.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
+                return np.empty(size + (0,), dtype=int)
+
+        # Heuristic rule for choosing the algorithm.
+        if total > 500000 or nsample > 12*ncolors:
```
What is the reason for this choice? Note that if it is changed the output sequence may also be changed.

charris commented:
We usually have a check for repeatability. Here, with the two different implementations, we need at least two. It isn't possible to try all the different color combinations, but maybe pick something simple and a long sequence. I used a hash of the sequences for some checks, but attention needs to be paid to little/big endian so that veiwed as bytes the sequences are the same. See line 179 in `test_random.py`.

charris commented:
For small sample lengths (< 256) you could just convert the output to `uint8` if you go the hash route.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4888,6 +4889,247 @@ cdef class RandomState:
+        ncolors = len(colors)
+        shape = _shape_from_size(size, ncolors)
+        samples = np.zeros(shape, dtype=np.int)
+        samples_ptr = <long*> PyArray_DATA(samples)
```
I think `np.int_` is the approved type for `long`.

charris commented:
This will likely have conflicts with #6938, which I want to merge first.

WarrenWeckesser commented:
> This will likely have conflicts with #6938, which I want to merge first.

No problem.  And thanks for the prompt "first look"!

charris commented:
I note that `rk_hypergeometric` is not implemented in the most efficient manner. It can't be completely fixed up in the obvious way and maintain backwards compatibility, but the `floor` function and cast to `long` can be gotten rid of.

WarrenWeckesser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -3249,7 +3249,6 @@ cdef class RandomState:
         but is sharper at the peak and has fatter tails. It represents the
         difference between two independent, identically distributed exponential
         random variables.
-
```
Oops, this blank line should not have been deleted.  I'll put it back when I update the pull request.

WarrenWeckesser commented:
In an inline comment, @charris asked about the rule used to choose between the two algorithms.  With `_multivariate_hypergeometric1` and `_multivariate_hypergeometric2` made public, I used this simple script (and other variations) to investigate the performance:

```
from __future__ import print_function, division

import sys
import numpy as np
import timeit

num_colors = [3, 5, 10, 15, 20, 25, 50, 100, 200, 500, 1000]
totals = [25, 50, 75, 100, 150, 200, 500, 750, 1000, 1500, 2000, 3000, 5000, 6000, 8000, 10000]

f = sys.stdout
for total in totals:
    for ncolors in num_colors:
        nsamples = np.unique((np.logspace(np.log10(3), np.log10(total), 25) + 0.5).astype(int))
        colors = np.bincount(np.random.choice(ncolors, total), minlength=ncolors)
        for nsample in nsamples:
            number = int(20000 / ncolors)
            t1 = timeit.timeit('np.random.multivariate_hypergeometric1(colors, nsample, size=100)',
                              globals=globals(), number=number)
            t2 = timeit.timeit('np.random.multivariate_hypergeometric2(colors, nsample, size=100)',
                              globals=globals(), number=number)
            if t1 < t2:
                best = 1
            else:
                best = 2
            print("%5d %5d %5d %10.6f %10.6f %d" %
                  (ncolors, total, nsample, t1, t2, best), file=f)
```
I found that algorithm 1 (iterated calls to `rk_hypergeometric`) was faster when `12*nsample > len(colors)`.  It is not a perfect rule, but in the cases where I saw it pick the wrong algorithm, the difference in performance was on the order of 25 to 30%.  (Note that in some extreme cases, the difference in performance can be several orders of magnitude.  The simple rule gets these right.)

I also experimented with even larger values of `sum(colors)` in an ipython shell.  When `sum(colors)` is huge (e.g. a million), algorithm 1 is faster.  So the rule that I ended up with is when `sum(colors) > 500000 or nsample > 12*len(colors)`, algorithm 1 is used.

I used only one computer for this experimentation, so I don't know how the result would be affected by using a different system.  I can reexpose the individual algorithms if anyone wants to experiment further.

I guess we could expose both algorithms permanently, with, say, a `method` argument, but I'd rather not do that without a compelling use-case.  If the rule for choosing the method works well enough, we shouldn't clutter the API with options that would rarely, if ever, be needed by a numpy user.  (If a compelling use-case shows up later, it is easy to add the option in a future version of numpy.)

WarrenWeckesser commented:
I pushed an update that addresses most of the comments so far.  I left the two algorithms as (private) python functions.  If anyone wants to experiment with the performance of the two algorithms, they are available as `numpy.random.mtrand._rand._multivariate_hypergeometric#`, where `#` is `1` or `2`.

wrwrwr commented:
Just an additional [data point](https://github.com/numpy/numpy/files/521407/callisto-u1604-35-at.txt) for the tuning (from a moderately aged Callisto processor running Python 3.5 on Ubuntu 16.04 with ATLAS).  I used [your script](https://github.com/numpy/numpy/files/521419/collect.py.txt) to gather the data and [this bit](https://github.com/numpy/numpy/files/521464/stats.py.txt) to process it.

For this data, the heuristic is correct in 96% of cases, that's not bad, though by changing 12 to 19 you can get close to 99%.

wrwrwr commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4612,6 +4614,300 @@ cdef class RandomState:
+                # `nsample` entries of `choices`.  The effect is
+                #     choices[:nsample] =
+                #         np.random.choice(total, size=nsample,
+                #                          replace=False)[:nsample]
```
Is the second `[:nsample]` needed in this example?

WarrenWeckesser replied:
No, I think that's left over from an earlier explanation.  I'll get rid of it.

wrwrwr commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4612,6 +4614,300 @@ cdef class RandomState:
+                #                          replace=False)[:nsample]
+                # while choices[nsample:] contains the rest of the permutation,
+                # but not fully randomized.
+                for j in range(0, nsample):
```
Just `range(nsample)`?

WarrenWeckesser replied:
Yup, thanks.

wrwrwr reviewed (commented):
Sorry for the nitpicks, feel free to ignore them, I haven't found anything more significant in the whole thing :-)

WarrenWeckesser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4612,6 +4614,300 @@ cdef class RandomState:
+           red_sample = rk_hypergeometric(ngood=nred, nbad=total - nred,
+                                          nsample=nsample)
+
+        The gives us the number of red marbles in the sample.  The number of
```
Oops.  Thanks.

WarrenWeckesser commented:
@wrwrwr All nits worth picking. :)  I'll fix them when I update again.  Thanks for testing the heuristic.   I'll experiment some more with 12 vs. 19 (or some other value) before updating the code.

wrwrwr commented:
Another bit. Couldn't we speed up cases with the number of samples close to the total number of items using symmetry (possibly just by adding a flag and some subtractions to algorithm 2)?

For example, `multivariate_hypergeometric(3 * [int(1e16)], int(3e16) - 5)` currently takes a while to compute, but could be sampled by choosing colors of the few items not to be drawn. The example is rather extreme, but my feeling is that such an approach could compete with algorithm 1 where it wins with 2.

homu commented:
:umbrella: The latest upstream changes (presumably #8219) made this pull request unmergeable. Please resolve the merge conflicts.

WarrenWeckesser commented:
> Couldn't we speed up cases with the number of samples close to the total number of items using symmetry?

Yes, great idea!  I pushed an updated version that takes advantage of the symmetry to improve the performance of both algorithms.  I also updated the condition where algorithm 1 is used to `min(nsample, sum(colors) - nsample) > 15*len(colors)`.

homu commented:
:umbrella: The latest upstream changes (presumably #5726) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #8795) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@WarrenWeckesser I fixed the conflict, so if you make further changes here you will want to pull down the version on your github repo, might need a force pull, (I was curious about the github option to fix conflicts)

charris commented:
@WarrenWeckesser Having a crossover point for the method still makes me nervous. Two other possibilities occur to me, the crossover point could be passed as argument with reasonable default value, or the method could be passed as an argument. The second rather appeals to me, especially if one of the methods is generally good, if not optimal, for common use cases.

charris commented:
@WarrenWeckesser I'm going to push this off to 1.14 unless you have time to work on it.

WarrenWeckesser commented:
No problem.  What's the status of https://github.com/numpy/numpy/pull/6938?  Earlier you said you'd like to get that merged before this.

charris commented:
@WarrenWeckesser I think #6938 needs a rethink and will comment there later. In any case, it doesn't  need to go in first and I think this PR is less problematic.

homu commented:
:umbrella: The latest upstream changes (presumably #9050) made this pull request unmergeable. Please resolve the merge conflicts.

WarrenWeckesser commented:
@charris wrote

> Having a crossover point for the method still makes me nervous. Two other possibilities occur to me, the crossover point could be passed as argument with reasonable default value, or the method could be passed as an argument. The second rather appeals to me, especially if one of the methods is generally good, if not optimal, for common use cases.

I went with the second option.  The method can be selected with the `method` argument.  That is simple to understand, and if in the future someone figures out a more reliable threshold, the option `method="auto"` can still be added (though not as the default).

Even if there are no suggested changes in the latest version, don't merge this yet.  I'm having second thoughts about calling one of the methods "shuffle".  There is a partial Fisher-Yates shuffle in the implementation, but I don't think that's a great name for the sampling method.

WarrenWeckesser commented:
I changed the name of the default method to `"count"`, so the two methods are `"count"` and `"iterative"`.

I'm now done until I get more feedback.

WarrenWeckesser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4640,6 +4642,368 @@ cdef class RandomState:
+        Draw samples from the multivariate hypergeometric distribution
+        ("count" algorithm).
+
+        `colors` must be a one-dimensional C-contiguous numpy array
```
How about using `cdef` to define the private methods, so they are not exposed as Python functions?  Then the only thing that calls them is the public function, and it already ensures that the array passed to the private methods is contiguous.  (I just tried it, and it works, but I haven't pushed the changed to this PR.)

charris replied:
That would work.

charris replied:
If you moved the lock out they wouldn't even need to be methods.

charris replied:
In fact, there is a lot of common code that could be moved out.

WarrenWeckesser replied:
True.  When I was trying to implement the automatic selection, it was convenient to keep the algorithms as accessible methods so I could bypass the automatic part and test the methods individually.  Now that there is a "method" argument, there is nothing lost by moving the two algorithms to functions outside of the `RandomState` class.  (Yeah, I realize we're talking about a suggestion that you made nine months ago. :)

charris replied:
> I realize we're talking about a suggestion that you made nine months ago

I'd completely forgotten. Must be getting to the age where I keep telling the same stories...

WarrenWeckesser replied:
As typically happens after I start writing nogil cython, I ended up ditching cython and writing the methods directly in C.  It also occurred to me that the "count" method could be simplified; the numpy version has a step that is not needed when the code is written using loops in cython or C.  I'll send an update when I have it all cleaned up and tested.

>  Must be getting to the age where I keep telling the same stories...

Nah, it just means your programming wisdom is consistent over time. :)

WarrenWeckesser commented:
I rewrote the two algorithms in C.  The "count" algorithm is also simpler.

Because of the size of this change, I added "WIP" to the PR title.

WarrenWeckesser commented:
I pushed an update, but the only change is to comments in the C code.

I need to clean up some C warnings related to comparing `size_t` and `long` values.

WarrenWeckesser commented:
I tweaked some declarations in the C to remove some compiler warnings.  This should be ready for review again.

WarrenWeckesser commented:
In the latest push, I tweaked the release notes to better follow the existing style.

charris commented:
@WarrenWeckesser Needs rebase, looks easy to fixup.

bashtage commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -123,6 +123,18 @@ cdef extern from "distributions.h":
+    # do not need the GIL, but they do need a lock on the state !!
+
+    int multivar_hypergeom_count(rk_state *state,
+                                 size_t num_colors, long *colors,
```
Since this isn't quite in, is it possible to use long long to avoid windows problems?

rkern replied:
FWIW, I'd recommend using an explicitly-sized integer type like `int64_t` instead of `long long` if we do decide to change anything here. I'm not sure that we ought to change the type in the signature, though. I'm guessing that inputs above `2**31` are unlikely and probably not going to work well with the algorithm regardless.

Internal variables holding intermediates might be worth upgrading to `int64_t`. I might recommend doing this kind of thing as a uniform change across all of the distributions, all at once, though.

bashtage replied:
On Windows `long` can be a pain since `int64` has become so common via pandas. I agree that explicit `int64` should be the default type for integer generators.

WarrenWeckesser replied:
The integers in, for example, `distributions.c` are all `long`, so I'm basically following the existing pattern.  If I had to justify it, I'd say that using `long` matches the default type of integer arrays in numpy (see, for example, Nathaniel's recent comment here: https://github.com/numpy/numpy/issues/9464#issuecomment-318185121).

Robert's suggestion to eventually update the types all at once sounds good.

rkern replied:
That was indeed the original motivation for using `long` in all of the other distributions' signatures.

bashtage replied:
I agree that this isn't the place to make those changes.  I do think accepting `int64` for integer datatypes is generally a good idea.

WarrenWeckesser commented:
In case anyone was thinking of merging this soon: wait a little longer.  I'm experimenting with a different algorithm.

charris commented:
Oh, just in time. I was starting a review with an eye towards merging.

WarrenWeckesser commented:
Nevermind, the alternative algorithm didn't provide an advantage over the existing two.  I've dropped the "WIP" from the title again.

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+*/
+
+int multivar_hypergeom_iterative(rk_state *state,
+                                 size_t num_colors, long *colors,
```
Is `size_t` the right thing here? The biggest we can get is `npy_intp`.

WarrenWeckesser replied:
Are there any guarantees about the bit width of `npy_intp` compared to C `long` (signed or unsigned)?  I'm not sure if these can be safely mixed in expressions and comparisons.

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+#include "distributions.h"
+
+/*
+    Draw samples from the multivariate hypergeometric distribution--
```
Multiline comments like
```
/*
 * blah
 */
```

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    i = 0;
+    while (i < num_sample) {
```
Why not a `for` loop?

charris replied:
Really, this is a classic `for` loop
```
for(i = 0; i < num_sample; i += num_colors) {
```

WarrenWeckesser replied:
Already fixed it in the update that I'm working on. :)  The only thing I have to do before sending the update is to address your comment about `size_t`.  I'll try to get this done later tonight (i.e. within the next 8 hours or so, for those not in my time zone).

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+}
+
+
+/*
```
See above in multiline comments.

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    i = 0;
+    while (i < num_sample) {
```
Use `for` loop?

charris commented on numpy/random/mtrand/mvhypergeom.c:
```diff
@@ -0,0 +1,246 @@
+    i = 0;
+    while (i < num_sample) {
+        /*
+         *  Fisher-Yates shuffle, but only loop through the first
```
Is Fisher-Yates implemented anywhere else? ISTR that there was some talk of using another algorithm for shuffle.

WarrenWeckesser replied:
This is basically a random choice without replacement.  Floyd's algorithm can be used (it is mentioned in some older numpy issues and pull requests).  In fact, that is one of the experiments I recently tried: Floyd's algorithm, using khash to implement the set of selected numbers.  However, my implementation didn't provide any clear advantage over the current two implementations, so I abandoned that approach.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -693,6 +693,97 @@ def test_multivariate_normal(self):
         assert_raises(ValueError, np.random.multivariate_normal, mean, cov,
                       check_valid='raise')
 
+    def test_multivariate_hypergeometric(self):
```
This is big enough that I'd rather split it up into separate tests in their own class. Note that you can use hashes for repeatability if you want to test larger numbers of samples.

charris replied:
Looks like a some trailing white space has crept into master, if that is fixable here that would be nice.

charris commented:
Some minor nits, otherwise looks good to me.

charris commented:
I'd like to put this in but am waiting for fixes of the style nits. @WarrenWeckesser To you have time to get to this?

WarrenWeckesser commented:
I have some scipy work that is ahead of this on my "to do" list.  And after reading comments in other numpy pull requests regarding `numpy.random.choice`, there is yet another algorithm I'd like to test first.  Can this wait a week?

charris commented:
No big rush, but I'm starting to look forward to the 1.14 release.

WarrenWeckesser commented:
@charris, I just realized that the code needs to check that the sum of `colors` does not exceed the largest long integer, taking possible overflow into account.  It is not difficult, but it will be another few days before I can make time to finish it off.

WarrenWeckesser commented:
FYI: While testing this code with some edge cases, I found a bug in `numpy.random.hypergeometric`.  It is not likely that anyone would encounter the bug during "normal" use, but it will be nice to get it fixed.  I've been working on that, and I'll have a pull request for it soon.

WarrenWeckesser commented:
> Long should be avoided when possible. For array sizes and indexing, use npy_intp. For error returns and such, use int.

That's what I'm doing in the update that I'm working on.  It is the values in the `colors` array that are longs.  This is consistent with the rest of the `mtrand` code.  (See the inline comments by @rkern and @bashtage on September 1.)

charris commented:
@WarrenWeckesser How is this looking?

WarrenWeckesser commented:
I updated the code to avoid potential overflow of the argument to `malloc`.  Without those checks, large values in `colors` could make the argument to `malloc` overflow.

I also added code to avoid the bug in the C function `rk_hypergeometric_hyp` that was reported in https://github.com/numpy/numpy/pull/9834.  The bug is avoided by simply not allowing arguments large enough to trigger the bug.

charris commented:
@WarrenWeckesser Do you consider this ready? Could use a rebase.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4653,6 +4673,191 @@ cdef class RandomState:
+                                                      num_colors, colors_ptr,
+                                                      nsample,
+                                                      sample_size, sample_ptr)
+
```
Might want to raise an error if method is not in `['count', 'iterative']`.

charris replied:
NVM.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4653,6 +4673,191 @@ cdef class RandomState:
+        For some cases, the 'iterative' algorithm may be significantly faster
+        than the 'count' algorithm.
+
+        .. versionadded:: 1.14.0
```
Actually like to put this up after the explanatory body.

charris replied:
Also, need an entry in `doc/source/reference/routines.random.rst`.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4653,6 +4673,191 @@ cdef class RandomState:
+
+        bad_colors = False
+        try:
+            colors = np.require(colors, dtype=np.int_, requirements='C')
```
This errors on my machine.

bashtage replied:
I don't think this works on Windows as expected:

```
colors = np.array([1,3,4,5,2**60],dtype=np.int64)

np.require(colors, dtype=np.int_, requirements='C')
Out[12]: array([1, 3, 4, 5, 0])
```

Is setting out-of-range values to 0 expected?

charris replied:
NVM, colors was already defined from matplotlib.

WarrenWeckesser replied:
@bashtage No, that's not the desired behavior.  A value that is too big to be handled correctly should result in a ValueError.

eric-wieser replied:
Note that that's just `colors.astype(np.int32)`, which shows the same behaviour. Casting out-of-range values is implemention/un-defined in C, and using the `% 2**32` behavior is not surprising.

bashtage replied:
I agree that it should raise.  I don't think require does raise though -- in the code above it silently converts.  In fact, 

```
np.require(np.array([np.pi]), dtype=np.int_, requirements='C')
Out[2]: array([3])
```

it also truncates floats.  Maybe you need `PyArray_FROM_OTF` like most other functions, which will raise TypeError iirc.

eric-wieser replied:
This would be far faster as `np.ascontiguousarray(colors, dtype=int)`

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4653,6 +4673,191 @@ cdef class RandomState:
+        >>> np.random.multivariate_hypergeometric(colors, 6, size=(2, 2))
+        array([[[4, 1, 1],
+                [4, 2, 0]],
+
```
The blank lines mess up the document rendering in the examples, best to remove them.

WarrenWeckesser commented:
Currently the code quietly casts floating point values to integers:

```
In [73]: colors = [np.nextafter(1, 0), 2]

In [74]: colors
Out[74]: [0.9999999999999999, 2]

In [75]: np.random.multivariate_hypergeometric(colors, 2, size=8)
Out[75]: 
array([[0, 2],
       [0, 2],
       [0, 2],
       [0, 2],
       [0, 2],
       [0, 2],
       [0, 2],
       [0, 2]])
```

What do you think of requiring the inputs to be integers?

charris commented:
I pushed this off to 1.15. I don't think it needs much more work, but it doesn't seem urgent.

WarrenWeckesser commented:
No problem.  Sorry I haven't been working on this more consistently.

charris commented:
@WarrenWeckesser  Where is this these days? I'd like to branch 1.15 soon.

WarrenWeckesser commented:
OK, I'll get back to it (again!).

WarrenWeckesser commented:
Has there been any progress on the "versioning" API for numpy.random?  It would be nice to know how that is going to work.  I've been experimenting with different implementations for this pull request, and I suspect that once it sees more use, we'll want to tweak the implementation.  (It would also be nice for https://github.com/numpy/numpy/pull/9834.)

bashtage commented:
@WarrenWeckesser while I think there is some momentum in getting a better, improvable random generator into NumPy, you should consider contributions to RandomState as one-off-chances that realistically will not be modifiable in the near future.

WarrenWeckesser commented:
I just pushed an update, but that was because I pushed everything when I updated https://github.com/numpy/numpy/pull/11138.  I have a few more changes to make (in particular, to work around https://github.com/numpy/numpy/pull/11138), so don't start reviewing this again yet.

charris commented:
Pushing off to 1.16.

WarrenWeckesser commented:
That's fine. I have ended up rewriting the hypergeometric distribution functions in private functions to be used in `multivariate_hypergeometric`.  It is more work, and it will take more effort to review, but it means that the multivariate hypergeometric function with two colors (which is the univariate hypergeometric distrbution) will be faster than the existing hypergeometric function.  When `numpy.random` gets an API for versioning, we'll be able to use the new code in `hypergeometric`.

WarrenWeckesser commented:
I have updated the code with a faster version of the "marginals" method.

This should not be merged yet.  I am investigating the behavior of the method when it is given extremely large arguments.

WarrenWeckesser commented:
The test failure on "Shippable" appears to be a bug that is not directly related to this pull request (thought possibly triggered because not all the tests set a random seed).  The test that is failing is `TestRandomDist.test_weibull_0`, which does `assert_equal(np.random.weibull(a=0), 0)`.  The problem is that, even in numpy 1.15.4, we have:

```
In [36]: np.__version__
Out[36]: '1.15.4'

In [37]: np.random.seed(123)

In [38]: np.random.weibull(a=0, size=10)
Out[38]: array([inf,  0.,  0.,  0., inf,  0., inf, inf,  0.,  0.])
```
Those `inf` values should not be there.  I'll create a new issue for this.

charris commented:
@WarrenWeckesser So are you happy with this in its current state?

WarrenWeckesser commented:
@charris, I just noticed that the code in the function `mvhg_marginals()` was making the substitution `nsample = total - nsample` if `nsample > total/2`, but then the routines that it (ultimately) calls, `hypergeometric_sample()` and `hypergeometric_hrua()`, also make a similar check and transformation for the (univariate) hypergeometric distribution.  I'm cleaning that up now.

WarrenWeckesser commented:
No problem.  I ran some tests for ranges of the inputs that I hadn't tried before, and I noticed some bias in the "big" tests that I've been running.  Now I need to determine if that issue is inherent in the HRUA algorithm, or if it is a limitation of (or bug in) my test code.

mattip commented:
@bashtage is this something we should consider before the 1.17 release, or can it be added afterward?

WarrenWeckesser commented:
Is there a target date for the release of 1.17?

bashtage commented:
As must this PR is only about multivariate_hypergeometric, it could wait until later.  However, I think @WarrenWeckesser is also planning on improving hypergeometric to work better in edge cases.  Any changes to hypergeometric really should get in before 1.17 since these will change the stream (and so might require a warning cycle).

WarrenWeckesser commented:
With randomgen now merged, it makes sense to break this pull request into two parts.  The first part is the rewrite of the univariate hypergeometric code.  That is now in https://github.com/numpy/numpy/pull/13761.  The second part will be to add the new multivariate hypergeometric distribution, but that might have to wait until 1.18.

charris commented:
@WarrenWeckesser Closing this then.

charris commented:
I assume there will be a new PR with the second part.

WarrenWeckesser commented:
@charris wrote

> I assume there will be a new PR with the second part.

Yes.  If the branching of 1.17 is postponed because of the ongoing discussion of the new randomgen code, it might be possible to get `multivariate_hypergeometric` into 1.17, but I'm not counting on that.
